### PR TITLE
Issue 3563 mdm nickname support

### PR DESCRIPTION
**Pull request by fil512**
Added search parameter modifier :nickname that can be used with 'name' or 'given' search parameters.
E.g. ?Patient?given:nickname=Kenny will match a patient with the given name Kenneth.  Also added MDM matching
algorithm named NICKNAME that matches based this.

### CODE REVIEW

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/nickname/NicknameInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+				// from theSearchParameterMap.keySet()
+				expandAnyNicknameParameters(paramName, orList);
+			}
+		}
+	}
+
+	/**
+	 * If a Parameter is a string parameter, and it has been set to expand Nicknames, perform the expansion.
+	 */
+	private void expandAnyNicknameParameters(String theParamName, List<IQueryParameterType> orList) {
```
Comment (oliviayou) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/nickname/NicknameInterceptor.java:42]: `theParamName` isn't used in this method.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleTarget.java [hunk 2]
```diff
@@ -48,11 +48,24 @@ private Map<String, String[]> stripMdmSuffix(Map<String, String[]> theParameters
...
+	private String stripMdmQualifier(String key) {
+		if (key.endsWith(Constants.PARAMQUALIFIER_MDM)) {
+			key = key.split(Constants.PARAMQUALIFIER_MDM)[0];
+		}
+		return key;
+	}
+
+	private String stripNicknameQualifier(String key) {
+		if (key.endsWith(Constants.PARAMQUALIFIER_NICKNAME)) {
+			key = key.split(Constants.PARAMQUALIFIER_NICKNAME)[0];
```
Comment (oliviayou) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleTarget.java:67]: nitpick: `theKey`

**Review by oliviayou — COMMENTED** [review]
This is neat! added some nitpicks

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/nickname/NicknameInterceptor.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/nickname/NicknameInterceptor.java:42]: thanks. added it to logging.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleTarget.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleTarget.java:67]: fixed

**Review by jmarchionatto — APPROVED** [review]

**Pull request merged**